### PR TITLE
Bugfix GCSToGCSOperator when copy an object without wildcard and exact_match=True

### PR DESCRIPTION
**Pull request by moiseenkov**
This PR fixes the following case.

The goal is to copy `source/foo.txt` to `dest/foo.txt` within a single GCS bucket.
1. Create a GCS bucket and upload two files to source directory like this:
```
gs://my-bucket/source/foo.txt
gs://my-bucket/source/foo.txt.abc
```
2. Upload the following DAG to a Cloud Composer environment:
```python
from airflow import DAG
from airflow.providers.google.cloud.transfers.gcs_to_gcs import GCSToGCSOperator
from datetime import datetime

with DAG(
    dag_id="gcs_to_gcs_fail_example",
    schedule_interval=None,
    catchup=False,
    start_date=datetime(2021,1,1)
) as dag:
    copy_file = GCSToGCSOperator(
        task_id="copy_file",
        source_bucket="my-bucket",
        source_object="source/foo.txt",
        destination_object="dest/foo.txt",
        exact_match=True,
    )
    copy_file
```
3. Run the DAG

**Expected bucket state:**
```
gs://my-bucket/source/foo.txt
gs://my-bucket/source/foo.txt.abc
gs://my-bucket/dest/foo.txt
```
**Actual (incorrect) bucket state:**
```
gs://my-bucket/source/foo.txt
gs://my-bucket/source/foo.txt.abc
gs://my-bucket/dest/foo.txt/source/foo.txt
```

======================================================

The reason for this bug was the lack of handling `exact_match=True` when objects are being copied without a wildcard. This problem is fixed in the current PR.

======================================================

However, if the flag is set to its default value `exact_match=False`, then the operator's result is different:
```
gs://my-bucket/source/foo.txt
gs://my-bucket/source/foo.txt.abc
gs://my-bucket/dest/foo.txt/source/foo.txt
gs://my-bucket/dest/foo.txt/source/foo.txt.abc
```
It's actually correct, because in general `source_object="path/to/the/file.txt"` is not treated as a file path, but as an object name **prefix** ([doc](https://cloud.google.com/storage/docs/samples/storage-list-files-with-prefix)). That's why the **prefix** `source_object="path/to/the/file.txt"` corresponds to both objects:
```
gs://my-bucket/source/foo.txt
gs://my-bucket/source/foo.txt.abc
```
And if the destination_object is set, then the destination object prefix is just built as a concatenation of the source prefix and the destination prefix. There is no difference for GCS what is being copied: a file or a folder - both of these entities are the same things - objects.

Perhaps, it makes sense to implement more "human friendly" logic, so the operator would act with inputs as with files and folders, but I think it should be another operator, because `GCSToGCSOperator`'s current implementation became too complicated for major changes. This is just my thoughts, I'm not insisting.

### CODE REVIEW

**Comment by eladkal** [comment]
> Bugfix for the following case (b/289486604).

What is that?

**Comment by moiseenkov** [comment]
> > Bugfix for the following case (b/289486604).
> 
> What is that?

Reference to our internal ticket. Removed it.

**Comment by eladkal** [comment]
> Reference to our internal ticket. Removed it.

Can you please edit title  and amend comment to a meaningful title?

**Comment by moiseenkov** [comment]
> > Reference to our internal ticket. Removed it.
> 
> Can you please edit title and amend comment to a meaningful title?

Sure! Done.

File: airflow/providers/google/cloud/transfers/gcs_to_gcs.py [hunk 1]
```diff
@@ -370,6 +370,9 @@ def _copy_source_without_wildcard(self, hook, prefix):
             self.source_bucket, prefix=prefix, delimiter=self.delimiter, match_glob=self.match_glob
         )
 
+        if self.exact_match:
```
Comment (uranusjr) [inline airflow/providers/google/cloud/transfers/gcs_to_gcs.py]: Why is `exact_match` needed to be checked here and again in `_check_exact_match`?

**Review by uranusjr — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Pull request merged**